hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/SparkHoodieMetadataBloomIndexHelper.java

```diff
@@ -0,0 +1,167 @@
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Helper for {@link HoodieBloomIndex} containing Spark-specific logic.
+ */
+public class SparkHoodieMetadataBloomIndexHelper extends MetadataBaseHoodieBloomIndexHelper {
+
+  private static final Logger LOG = LogManager.getLogger(SparkHoodieMetadataBloomIndexHelper.class);
+  private static final SparkHoodieMetadataBloomIndexHelper SINGLETON_INSTANCE =
+      new SparkHoodieMetadataBloomIndexHelper();
+
+  private SparkHoodieMetadataBloomIndexHelper() {
+  }
+
+  public static SparkHoodieMetadataBloomIndexHelper getInstance() {
+    return SINGLETON_INSTANCE;
+  }
+
+  protected HoodieData<Pair<String, IndexFileFilter>> buildPartitionIndexFileFilter(
+      HoodieWriteConfig config, final HoodieEngineContext context, final HoodieTable hoodieTable,
+      HoodiePairData<String, String> partitionRecordKeyPairs) {
+    final String keyField = hoodieTable.getMetaClient().getTableConfig().getRecordKeyFieldProp();
+    final int inputParallelism = HoodieJavaPairRDD.getJavaPairRDD(partitionRecordKeyPairs).partitions().size();
+    final int joinParallelism = Math.max(inputParallelism, config.getBloomIndexParallelism());
+    LOG.info("Building partition index file filter, input parallelism: " + inputParallelism
+        + ", bloom index parallelism: " + config.getBloomIndexParallelism());
+    JavaRDD<String> partitionsRDD = HoodieJavaPairRDD.getJavaPairRDD(partitionRecordKeyPairs).keys().distinct().repartition(joinParallelism);
+
+    JavaRDD<Pair<String, IndexFileFilter>> partitionIndexFileFilterRDD = partitionsRDD.mapPartitionsWithIndex(
+        new Function2<Integer, Iterator<String>, Iterator<Pair<String, IndexFileFilter>>>() {
+          @Override
+          public Iterator<Pair<String, IndexFileFilter>> call(Integer integer, Iterator<String> partitionItr) throws Exception {
+            HoodieTimer timer = new HoodieTimer().startTimer();
+            final List<Pair<String, IndexFileFilter>> result = new ArrayList<>();
+            final List<Pair<String, String>> columnStatKeys = new ArrayList<>();
+            long partitionCount = 0;
+            while (partitionItr.hasNext()) {
+              String partitionName = partitionItr.next();
+              List<String> partitionFileNameList = HoodieIndexUtils.getLatestBaseFilesForPartition(partitionName,
+                      hoodieTable).stream().map(baseFile -> baseFile.getFileName())
+                  .collect(toList());
+              partitionCount++;
+
+              if (partitionFileNameList.isEmpty()) {
+                result.add(new ImmutablePair<String, IndexFileFilter>(partitionName,
+                    new IntervalTreeBasedIndexFileFilter(partitionName, Collections.emptyList())));
```

Comment:
is it a conscious decision to not add ListBasedIndexFileFilter support in this index.